scripts/builtin/xgboost.dml

```diff
@@ -0,0 +1,780 @@
+
+      current_node = matrix(0, rows=6, cols=1)
+      current_node[1,1] = node[1,1] # node id
+      current_node[2,1] = tree_id
+      current_node[3,1] = offset # offset to left child
+      current_node[4,1] = used_col # features used
+
+      if(used_col == 0.0) { # is a single value or max depth reached => leaf
+         current_node[5,1] = 0 #  0 if leaf node
+         assert(threshold == 0)
+         current_node[6,1] = output_value
+      }
+      else if (as.scalar(R[1, used_col]) == 2.0) {
+        current_node[5, 1] = 2.0 # categorical
+        assert(threshold == 0.0) # should be 0 at categorical
+        current_node[6,1] = threshold
+      }
+      else {
+        current_node[5,1] = 1.0 # 1 if it is scalar
+        current_node[6,1] = threshold # split value ( 0 if leaf or categorical)
+      }
+
+      if (ncol(M) == 0 & nrow(M) == 0) { # first node
+        new_M = current_node
+      } else {
+        new_M = cbind(M, current_node)
+      }
+}
+
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:   R: Location to read the matrix R(nx1) which for each feature in X contains the following information
+#					- R[,2]: 1 (scalar feature)
+#			    	- R[,1]: 2 (categorical feature)
+# INPUT:    col: the col that is desired to know the type
+# OUTPUT:   [1.0,scalar] or  [2.0,categorical]
+#-----------------------------------------------------------------------------------------------------------------------
+getTypeOfFeature = function(Matrix[Double] R, Double col)  return(Double type) {
+  type = as.scalar(R[1, col])
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    X: a nxn matrix (X Matrix from input)
+# INPUT:    y: a 1xn matrix (y Matrix from input)
+# INPUT:    prediction: a 1xn matrix with the current predictions for y (median at first tree)
+# INPUT:    vector: a 1xn matrix (contains 1 for relevant and 0 for non-relevant value)
+# OUTPUT:   new_X: the new X containing only the relevant rows
+# OUTPUT:   new_y: the new y containing only the relevant rows
+# OUTPUT:   new_X_full: the new X containing the relevant rows + 0 for non_relevant
+# OUTPUT:   new_prediction: the current prediction containing only the relevant rows
+updateMatrices = function(Matrix[Double] X, Matrix[Double] y, Matrix[Double] prediction, Matrix[Double] vector)
+  return(Matrix[Double] new_X, Matrix[Double] new_y, Matrix[Double] new_X_full, Matrix[Double] new_prediction) {
+
+  assert(ncol(vector) == 1)
+  assert(nrow(vector) == nrow(X))
+
+  new_X = matrix(0, rows = 0, cols = 0)
+  new_y = matrix(0, rows = 0, cols = 0)
+  zero_vec = matrix(0, rows=nrow(vector), cols=ncol(X))
+  nan_vec = fillMatrixWithNaN(zero_vec) # could not directly set to NaN
+  new_X_full = matrix(0, rows = 0, cols=0)
+  new_prediction = matrix(0, rows=0, cols=0)
+  for(i in 1:nrow(vector)) {
```

Comment:
This here is very slow, taking one row at a time and r binding it to the output.
If i remember correctly  @mboehm7 , have some magic selection algorithm,
where you make a boolean matrix, and slice out rows based on that.

Could i get a comment @mboehm7